### PR TITLE
[WIP] add fix-permissions

### PR DESCRIPTION
**Pull request by xu-cheng**
Another take of #45842.

cc @mikemcquaid @bfontaine  @DomT4 @UniqMartin

### CODE REVIEW

File: Library/Homebrew/cmd/fix-permissions.rb [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        opoo "#{HOMEBREW_CACHE} doesn't exist."
+      end
+
+      puts "To fix permissions, run `brew fix-permissions --force`"
+      Homebrew.failed = true
+    end
+  end
+end
+
+class PermissionChecker
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/fix-permissions.rb]: This class will be copied to install script.

File: Library/brew.rb [hunk 2]
```diff
@@ -98,7 +98,7 @@ def require?(path)
   cmd = HOMEBREW_INTERNAL_COMMAND_ALIASES.fetch(cmd, cmd)
 
   sudo_check = %w[ install reinstall postinstall link pin unpin
-                   update upgrade create migrate tap switch ]
+                   update upgrade create migrate tap switch fix-permissions ]
```
Comment (mikemcquaid) [inline Library/brew.rb]: Seems like maybe we should allow this and just default to `--force` if you're using `sudo`?

**Comment by mikemcquaid** [comment]
Seems pretty reasonable. I'd like to see how the `install` script looks as presumably the YAML will need embedded in that script too. Maybe also worth doing the `git checkout -f master; git reset --hard origin/master` too.

**Comment by DomT4** [comment]
This looks great Xu. Moving away from the constant repetition of `sudo chown -R blah` in `doctor` is a very welcome fix.

File: Library/Homebrew/cmd/fix-permissions.rb [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    load_from_yaml (HOMEBREW_LIBRARY_PATH/"permission_checklist.yml").read
+  end
+
+  # only used in install script
+  def self.load_from_remote
+    url = "https://raw.githubusercontent.com/Homebrew/homebrew/master/Library/Homebrew/permission_checklist.yml"
+    remote_yaml = `curl -fsSL ${url}`
+    raise "Fail to fetch remote resource: #{url}" if remote_yaml.empty?
+    load_from_yaml remote_yaml
+  end
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/fix-permissions.rb]: > I'd like to see how the install script looks as presumably the YAML will need embedded in that script too. 

@mikemcquaid This is how install script will load yaml file. So we only need to maintain one checklist.

File: Library/brew.rb (hunk 2 above)
Comment (xu-cheng) [inline Library/brew.rb]: I'm not sure whether running `sudo brew fix-permissions` would cause any user mismatch problem(especially in multiuser system).

Comment (mikemcquaid) [inline Library/brew.rb]: Ok. Might be worth checking and/or having a more specific message?

Comment (xu-cheng) [inline Library/brew.rb]: Did some local tests. I think current `sudo_check` behavior is exactly what we want.
```
$ ruby -e 'puts File.owned?("/usr/local/bin/brew")'
true

$ sudo ruby -e 'puts File.owned?("/usr/local/bin/brew")'
false

$ ruby -e 'puts ENV["USER"]'
xucheng

$ sudo ruby -e 'puts ENV["USER"]'
root
```

Comment (mikemcquaid) [inline Library/brew.rb]: :+1:

**Comment by xu-cheng** [comment]
I completely rewrote this PR. tests are required.

Remaining TODO list:
* [ ] write test case.
* [ ] test this against all travis images to check possible false positive result during `brew update`.

**Comment by mikemcquaid** [comment]
Looks good. Would it be worth adding unification with `install` to the TODO list? With `brew update` we could use the `fetch` and `merge`/`rebase` separation to check the permissions for the actually updated files.

**Comment by apjanke** [comment]
This looks like something we should still pull in. Want to migrate it to homebrew/brew to keep the discussion live?

**Comment by xu-cheng** [comment]
This is still on my TODO list.

**Comment by ilovezfs** [comment]
I'm not thrilled with the idea of having a built-in command that runs sudo. Maybe this should be an external utility the user has to download separately like the install script?

**Comment by mikemcquaid** [comment]
@ilovezfs I think it's fine if we signal loudly enough and (for a change) perhaps prompt for confirmation.

**Comment by mikemcquaid** [comment]
@xu-cheng I think it would be good to have this sooner rather than later (e.g. pre-1.0) so could you consider either passing this on to someone else or finishing it up? Thanks.

**Comment by xu-cheng** [comment]
I'm kinda busy on other matters. So if anyone want to work on this, it would be fine with me. Otherwise, it will just be on my TODO list for the time being.

**Comment by mikemcquaid** [comment]
@xu-cheng Thanks!

**Comment by MikeMcQuaid** [comment]
Can this be ported over to the Homebrew/brew repo when it's ready to 🚢? Thanks!

**Pull request closed without merging**